internal/gardenclient/shoot_client.go

```diff
@@ -0,0 +1,273 @@
+/*
+SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package gardenclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/Masterminds/semver"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	clientauthenticationv1beta1 "k8s.io/client-go/pkg/apis/clientauthentication/v1beta1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdlatest "k8s.io/client-go/tools/clientcmd/api/latest"
+	clientcmdv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+const (
+	// ShootProjectSecretSuffixCACluster is a constant for a shoot project secret with suffix 'ca-cluster'.
+	ShootProjectSecretSuffixCACluster = "ca-cluster"
+	// DataKeyCertificateCA is the key in a secret data holding the CA certificate.
+	DataKeyCertificateCA = "ca.crt"
+)
+
+// shootKubeconfigRequest is a struct which holds information about a Kubeconfig to be generated.
+type shootKubeconfigRequest struct {
+	// cluster holds all the cluster on which the kube-apiserver can be reached
+	clusters []cluster
+	// namespace is the namespace where the shoot resides
+	namespace string
+	// shootName is the name of the shoot
+	shootName string
+	// gardenClusterIdentity is the cluster identifier of the garden cluster.
+	gardenClusterIdentity string
+}
+
+// cluster holds the data to describe and connect to a kubernetes cluster
+type cluster struct {
+	// name is the name of the shoot advertised address, usually "external", "internal" or "unmanaged"
+	name string
+	// apiServerHost is the host of the kube-apiserver
+	apiServerHost string
+
+	// caCert holds the ca certificate for the cluster
+	//+optional
+	caCert []byte
+}
+
+// ExecPluginConfig contains a reference to the garden and shoot cluster
+type ExecPluginConfig struct {
+	// ShootRef references the shoot cluster
+	ShootRef ShootRef `json:"shootRef"`
+	// GardenClusterIdentity is the cluster identifier of the garden cluster.
+	// See cluster-identity ConfigMap in kube-system namespace of the garden cluster
+	GardenClusterIdentity string `json:"gardenClusterIdentity"`
+}
+
```

Comment:
```suggestion

func (e ExecPluginConfig) GetObjectKind() schema.ObjectKind {
	return schema.EmptyObjectKind
}

func (e ExecPluginConfig) DeepCopyObject() runtime.Object {
	return ExecPluginConfig{
		ShootRef: ShootRef{
			Namespace: e.ShootRef.Namespace,
			Name:      e.ShootRef.Name,
		},
		GardenClusterIdentity: e.GardenClusterIdentity,
	}
}
```
We need to implement `runtime.Object` interface to use clientcmdapi.

Reply:
✅